In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
df = pd.read_csv(r'C:\Users\nehat\OneDrive\Desktop\CF DA Immersion\Achievement 4\Instacart Basket Analysis 09\02 Data\Original Data\4.3_orders_products\orders.csv', index_col = False)

In [3]:
df_ords = pd.read_csv(r'C:\Users\nehat\OneDrive\Desktop\CF DA Immersion\Achievement 4\Instacart Basket Analysis 09\02 Data\Original Data\4.3_orders_products\orders.csv', index_col = False)

In [4]:
a 

In [40]:
path  = r'C:\Users\nehat\OneDrive\Desktop\CF DA Immersion\Achievement 4\Instacart Basket Analysis 09'

In [5]:
# Create a datafrome

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [46]:
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# check for missing values 
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#view the 16 values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
#drop the 16 values
df_prods_clean.shape

(49677, 5)

In [21]:
#look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
#duplicates found
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
# check number of rows
df_prods_clean.shape

(49677, 5)

In [24]:
#create dataframe that doesnt include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

In [41]:
# export dataframe to prepared data
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

## TASK 4.5

In [ ]:
#1 consistency checks above for df_prods

In [39]:
#2 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The data for 'prices' seems off and should be investigated further. With a minimum of '1' and a maximum of '99999' the quartile results are on the lower side, which makes me think there is something wrong.


In [45]:
#3 Check for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

There seems to be no mixed-type data in the df_ords dataframe.

In [47]:
#5 Check for missing values

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [53]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [54]:
df_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

When I first ran the code to check for missing values 'days_since_prior_order' shows 206209 missing values, so to look in to it further I ran a subset to show the rows with the missing values.

In [ ]:
#6 addressing the missing values

In [56]:
# number of rows and columns in the dataframe

df_ords.shape

(3421083, 7)

In [58]:
# find the mean

df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [62]:
# replace missing values with the mean value given in the table above for days_since_prior_order

df_ords['days_since_prior_order'].fillna(1.111484 , inplace=True)

In [68]:
# to check if the mean value replaced 'NaN'
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                1.111484  
1                        7               15.000000  
2                       12               21.000000  
3                        7               29.000000  
4                       15               28.000000  
...                    ...                     ...  
3421078                 18               29.000000  
3421079                 10               30.000000  
3421080                 12               18.000000  
3421081                 12                7.000000  
3421082                 14               30.000000  

[3421083 rows x 7 columns]

I decided to use the mean value of the dataframe as there are too many values missing and to drop all those columns may not be a good decision to make. As I am imputing in replacement for a numeric value it made sense to me to use the average. 

In [72]:
#7 Duplicate values

df_dup = df_ords[df_ords.duplicated()]

In [73]:
df_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

no duplicates found


#8 no duplicates to address

In [74]:
# export dataframe to prepared data

df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))
